src/main/scala/tech/cryptonomic/conseil/tezos/TezosNodeOperator.scala

```diff
@@ -151,102 +158,68 @@ class TezosNodeOperator(node: TezosRPCInterface)(implicit ec: ExecutionContext)
+    val blocksFromFork = followFork match {
+      case false => Future.successful(List.empty : List[Block])
+      case true => Future.successful(List.empty : List[Block])
+    }
+    for {
+      forkBlocks <- blocksFromFork
+      rangeBlocks <- blocksInRange
+    } yield forkBlocks ++ rangeBlocks
   }
 
   /**
-    * Traverses Tezos blockchain as parameterized, without waiting for the answer
-    *
-    * @param network       Which Tezos network to go against
-    * @param startingBlock Current block to start from
-    * @param minLevel      Minimum level at which to stop
-    * @param maxLevel      Level at which to start collecting blocks
-    * @param followFork    If the predecessor of the minLevel block appears to be on a fork, also capture the blocks on the fork.
-    * @return              All collected blocks, wrapped in a [[Future]]
+    * Gets block from Tezos Blockchains, as well as their associated operation, from minLevel to maxLevel.
+    * @param network Which Tezos network to go against
+    * @param hash Hash of block at max level.
+    * @param minLevel Minimum level, at which we stop.
+    * @param maxLevel Level at which to stop collecting blocks.
+    * @return
     */
-  private def fetchBoundedBlockChain(
-    network: String,
-    startingBlock: Block,
-    minLevel: Int,
-    maxLevel: Int,
-    followFork: Boolean
-  ): Future[List[Block]] = {
+  private def processBlocks(
+                           network : String,
+                           hash: String,
+                           minLevel: Int,
+                           maxLevel: Int
+                           ): Future[List[Block]] = {
+    val maxOffset: Int = maxLevel - minLevel
+    val offsets = (0 to maxOffset).toList.map(_.toString)
+    val blockMetadataUrls = offsets.map{off => s"blocks/$hash~$off"}
+
+    val jsonToBlockMetadata: String => BlockMetadata =
+      json => fromJson[BlockMetadata](json)
+
     val jsonToOperationGroups: String => List[OperationGroup] =
       json => fromJson[List[List[OperationGroup]]](json).flatten
 
     for {
-      metadataChain <- fetchChainedMetadata(network, startingBlock.metadata, minLevel, maxLevel, followFork)
-      operationsUrls = metadataChain.map(d => s"blocks/${d.hash}/operations")
-      operations <- node.runBatchedGetQuery(network, operationsUrls, blockOperationsFetchConcurrency) map (all => all.map(jsonToOperationGroups))
-    } yield metadataChain.zip(operations).map(Block.tupled)
+      blockMetadata <- node.runBatchedGetQuery(network, blockMetadataUrls, blockOperationsFetchConcurrency) map (all => all.map(jsonToBlockMetadata))
+      blockOperationUrls = blockMetadata.map(metadata => s"blocks/${metadata.hash}/operations")
+      blockOperations <- node.runBatchedGetQuery(network, blockOperationUrls, blockOperationsFetchConcurrency) map (all => all.map(jsonToOperationGroups))
+    } yield blockMetadata.zip(blockOperations).map(Block.tupled)
```

Comment:
@ba205 since you asked I go for a longer answer here for: "Are we guaranteed that the `zip` operation correctly matches the operations with the corresponding metadata?"

In general we have no guarantees unless the `runBatchedGetQuery` adds those (guarantees) and documents them in the method API.

Right now it works correctly because the implementation of the batched call uses a stream that transforms each input url to the corresponding operation result, and doesn't re-order the stream elements (i.e. we use `mapAsync` instead of `mapAsyncUnordered`).

To be more robust to future changes and refactors we should add a visible marker in the implementation, like comments about the expected invariants of the call...
or returning some identifier with the operation get result that allows to match it with the proper metadata

Reply:
Okay, added documentation to address this concern.